lib/common/xml.c

```diff
@@ -618,6 +618,40 @@ expand_plus_plus(xmlNode * target, const char *name, const char *value)
     return;
 }
 
+/*!
+ * \internal
+ * \brief Remove an XML element's attributes that match some criteria
+ *
+ * \param[in,out] element    XML element to modify
+ * \param[in]     match      If not NULL, only remove attributes for which
+ *                           this function returns true
+ * \param[in]     user_data  Data to pass to \p match
+ */
+void
+pcmk__xe_remove_matching_attrs(xmlNode *element,
+                               bool (*match)(xmlAttrPtr, void *),
+                               void *user_data)
+{
+    xmlAttrPtr next = NULL;
+
+    for (xmlAttrPtr a = pcmk__xe_first_attr(element); a != NULL; a = next) {
+        next = a->next; // Grab now because attribute might get removed
+        if ((match == NULL) || match(a, user_data)) {
+            if (!pcmk__check_acl(element, NULL, xpf_acl_write)) {
+                crm_trace("ACLs prevent removal of %s attribute from %s element",
+                          (const char *) a->name, (const char *) element->name);
+
```

Comment:
good question is what is better here.
like this we cycle through the attributes even if we aren't allowed to remove them anyway but we get a nice list of what exactly couldn't be removed but we don't get a trace of unauthorized deletion attempts in case nothing matching was found.
if we bail out early we might save cycles and would trace any unauthorized attempt.

Reply:
Hmm, not sure. I think if nothing is actually deleted, then there's nothing wrong with trying (the attempt might be an internal side effect of some other acceptable change). But it does make sense to me to bail out the first time an attribute is blocked by ACLs, since we only allow element-level and not attribute-level blocking. I'll add a separate commit for that.

Reply:
It's also interesting that the only consequence of a block is a trace message, which makes me think that attempting to delete nothing is unlikely to be interesting.

Also, it looks like the callers of the new function for the most part aren't using it on the CIB, so ACLs won't actually be involved.

Reply:
Well wasn't sure - just thought it might be useful to be pointed at something that is probably a result of wrong usage of interfaces.